### ISSUE TITLE
Verify/fix Ubuntu 24.04 compatibility

### ISSUE THREAD

**Issue by friday** [Ulauncher/Ulauncher#1357]
### Environment

Distro: Mint 22 (Ubuntu 24.04) DE: Cinnamon 6.0.4

### Bug description

From #1355:

1:
I tried `make deb` and then tried running the ulauncher .deb file but it gave the error:

    ./ulauncher_6.0.0~beta4_all.deb: line 1: syntax error near unexpected token `newline'
    ./ulauncher_6.0.0~beta4_all.deb: line 1: `!<arch>'

2:

When I download the .deb and double click it:

![image](https://github.com/Ulauncher/Ulauncher/assets/16354856/465ca962-f608-4855-88b6-06a6cfbb7786)

```sh
$ apt install python3-levenshtein
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Package python3-levenshtein is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'python3-levenshtein' has no installation candidate
```

I guess `./ulauncher_6.0.0-beta4_all.deb` doesn't work for installing .deb files? That's what I tried in the comment above.  My bad.  I tried this instead:

```sh
$ sudo dpkg -i ulauncher_6.0.0~a100-0ubuntu1ppa1~noble_all.deb 
dpkg: warning: downgrading ulauncher from 6.0.0~beta4-0ubuntu1 to 6.0.0~a100-0ubuntu1ppa1~noble
(Reading database ... 530500 files and directories currently installed.)
Preparing to unpack ulauncher_6.0.0~a100-0ubuntu1ppa1~noble_all.deb ...
Unpacking ulauncher (6.0.0~a100-0ubuntu1ppa1~noble) over (6.0.0~beta4-0ubuntu1) ...
dpkg: dependency problems prevent configuration of ulauncher:
 ulauncher depends on python3-levenshtein; however:
  Package python3-levenshtein is not installed.

dpkg: error processing package ulauncher (--install):
dependency problems - leaving unconfigured
Processing triggers for gnome-menus (3.36.0-1.1ubuntu1) ...
Processing triggers for mate-menus (1.26.1-1) ...
Processing triggers for desktop-file-utils (0.26+mint4) ...
Processing triggers for mailcap (3.70+nmu1ubuntu1) ...
Processing triggers for gnome-icon-theme (3.12.0-5) ...
Processing triggers for hicolor-icon-theme (0.17-2) ...
Errors were encountered while processing:
 ulauncher
```

You can see here the dependency was removed: https://answers.launchpad.net/ubuntu/noble/amd64/python3-levenshtein

### Communication guidelines

- [X] I have read and agree with Ulauncher's communication guidelines and checked the troubleshooting page for self-help.
- [X] I have tried to understand the error message and help myself, before reporting it.
- [X] My issue description has clear steps for how to reproduce the issue.
- [X] This issue applies to the whole Ulauncher application, not for a specific extension.

### Ulauncher v6

- [X] I have made sure the bug isn't already fixed in v6.

**Comment by friday** [comment]
@ddouglas87 The [latest beta release](https://github.com/Ulauncher/Ulauncher/releases/tag/v6.0.0-beta5) now includes ubuntu 24.04 in the PPA, and it has python3-levenshtein as a recommended dependency instead of required. It should be possible to install with the [regular instructions](https://ulauncher.io/#Download) (for the dev build). Please let me know if not. 

The build issue will have to wait. It may be an issue with the build packages in 24.04. I tried setting up a docker image with Ubuntu 24.04 to replicate it, but the required packages didn't even install.

**Comment by oklopfer** [comment]
Just tested the beta5 release and unfortunately the deb still treats Levenshtein as a dependency instead of recommended, so it still does not install correctly on 24.04.

**Comment by friday** [comment]
Thank you @oklopfer. I guess recommended dependencies are actually treated as normal dependencies then.

There seems to be flags to ignore recommended dependencies https://askubuntu.com/a/887704

This of course is not a solution for Ulauncher, but possibly a workaround for you in the meanwhile.

**Comment by oklopfer** [comment]
Built locally and was able to fix it by removing this line: https://github.com/Ulauncher/Ulauncher/blob/2eb7768c47fc25a5eb4f85032512ca9280952825/setup.cfg#L41

now I have:
```bash
$ sudo apt install ./dist/ulauncher_6.0.0~beta5_all.deb
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Note, selecting 'ulauncher' instead of './dist/ulauncher_6.0.0~beta5_all.deb'
The following packages were automatically installed and are no longer required:
  python3-cliapp python3-markdown python3-ttystatus python3-yaml
Use 'sudo apt autoremove' to remove them.
The following additional packages will be installed:
  [...]
Recommended packages:
  python3-levenshtein
The following NEW packages will be installed:
  [...]
  ```

**Comment by oklopfer** [comment]
> Thank you @oklopfer. I guess recommended dependencies are actually treated as normal dependencies then.
> 
> There seems to be flags to ignore recommended dependencies https://askubuntu.com/a/887704
> 
> This of course is not a solution for Ulauncher, but possibly a workaround for you in the meanwhile.

You responded a second too quick, I've got a soln!

**Comment by friday** [comment]
Thank you. I will remove that and make a new release (but the same version number likely) soon.

**Comment by friday** [comment]
@oklopfer I have now pushed that change and updated the deb in the release to a new build https://github.com/Ulauncher/Ulauncher/releases/tag/v6.0.0-beta5 (hopefully nobody grabbed the checksums for the old one :sweat_smile:)

I also updated the build system to pep-517, but I think it should work the same for pacstall.

More info on why we removed the legacy keybinder: https://github.com/Ulauncher/Ulauncher/pull/1256 (not to mention all the bug reports and questions we got due to not working on Wayland, and the fact that this makes it possible to eventually implement #1304).

**Comment by oklopfer** [comment]
install from the new deb works perfectly - thanks for your work and an awesome project!

**Comment by friday** [comment]
Also just pushed a new release for the PPA just for Noble (currently building) https://launchpad.net/~agornostal/+archive/ubuntu/ulauncher-dev/+packages

That won't affect you if you use pacstall of course (I never tried it, but I'm on Arch, so I definitely get the appeal).

**Comment by friday** [comment]
@oklopfer I was curious and checked your PR.

I also noticed the dependencies are outdated for v6. These are our required dependencies now (all of them were required for v5 too):
* python3-all
* python3-gi
* gobject-introspection
* gir1.2-glib-2.0
* gir1.2-gtk-3.0
* gir1.2-webkit2-4.0

These are optional (all new for v6, except AyatanaAppIndicator3):
* gir1.2-gtklayershell-0.1 (for Wayland support, see https://github.com/Ulauncher/Ulauncher/discussions/1353)
* gir1.2-wnck-3.0 (for https://github.com/Ulauncher/Ulauncher/pull/979)
* XApp or AyatanaAppIndicator3 (for tray icon support, you only need one of them and XApp is the best  option since it can distinguish between left and right clicks)

**Comment by friday** [comment]
Closing the issue. The syntax error sounds like an issue with the Ubuntu 24.04 packages for building python packages, since it's still in 2 months until the stable release. So I will not focus on that now unless more people encounter/report it.

**Comment by oklopfer** [comment]
updated those dependencies now, thanks again :P

**Comment by ddouglas87** [comment]
Just tried it.  Works like a charm.

**Comment by oklopfer** [comment]
Hey @friday, I hate to re-open this issue, but it seems there's a new dependency issue with 24.04 (and Sid too in the near future). The package that's causing the issue is `gir1.2-webkit2-4.0`; I've done some investigating to see what's up, and I have definitive conclusions. The package exists-ish on 24.04, in that it is in every other supported architecture *except* for amd64:
<img width="302" alt="Screenshot 2024-04-11 at 1 31 05 PM" src="https://github.com/Ulauncher/Ulauncher/assets/104327997/39c83a68-2778-468e-9cca-3f550acb1c21">
https://packages.ubuntu.com/noble/gir1.2-webkit2-4.0

Looking at their launchpad, we can see it was actually deleted from the disk, and they note it due to the `xz` backdoor, which is a bit peculiar:
<img width="1451" alt="Screenshot 2024-04-11 at 1 33 18 PM" src="https://github.com/Ulauncher/Ulauncher/assets/104327997/824e744f-b65f-41ce-a43b-7015e7a3e6cc">
https://launchpad.net/ubuntu/noble/amd64/gir1.2-webkit2-4.0


Interestingly, though, the version here is `2.43.3-1build2`, while the ones in Sid are at `2.42.5-2`:
<img width="619" alt="Screenshot 2024-04-11 at 1 35 45 PM" src="https://github.com/Ulauncher/Ulauncher/assets/104327997/8ce3eda7-dbed-417a-9f01-06b92a167f13">
https://packages.debian.org/sid/gir1.2-webkit2-4.0

But even more interesting to me, is that both Sid + Noble/Devel/24.04 have all of the other ones updates to `2.44.1-1`, even higher:
<img width="650" alt="Screenshot 2024-04-11 at 1 37 56 PM" src="https://github.com/Ulauncher/Ulauncher/assets/104327997/8d5cea93-c1ea-4499-ada1-368775f49525">
https://packages.debian.org/sid/gir1.2-webkit2-4.1
<img width="558" alt="Screenshot 2024-04-11 at 1 38 20 PM" src="https://github.com/Ulauncher/Ulauncher/assets/104327997/14821dbd-4f27-4056-b009-52e4dcf23e3b">
<img width="557" alt="Screenshot 2024-04-11 at 1 41 02 PM" src="https://github.com/Ulauncher/Ulauncher/assets/104327997/7054e2ce-443b-4581-ba6c-8f8d04f56b8a">
https://launchpad.net/ubuntu/+source/webkit2gtk

And looking at the changelog, sure enough, this is the case: 
<img width="592" alt="Screenshot 2024-04-11 at 1 43 20 PM" src="https://github.com/Ulauncher/Ulauncher/assets/104327997/bff580fa-4f5e-4ee0-9964-58efb9ef927b">
https://metadata.ftp-master.debian.org/changelogs//main/w/webkit2gtk/webkit2gtk_2.44.1-1_changelog

Meaning it is now a deprecated package and is going to disappear soon, and already has begun to for Ubuntu. `gir1.2-webkit2-4.1` seems to be extremely similar to `gir1.2-webkit2-4.0`, but with a very key difference of bumping up to `libsoup v3` from `libsoup v2`. I'm not sure what kind of changes you might need to make for this, but I hope I've provided enough to get you started. Thanks again :)

**Comment by friday** [comment]
> The package exists-ish on 24.04, in that it is in every other supported architecture except for amd64:

That's their distros problem that they created, and only they can fix. They will have to. There are probably hundreds of apps depending on this, including many core Gnome Desktop components, [Evolution](https://gitlab.gnome.org/GNOME/evolution/-/wikis/home), Geary, Lutris and many more.

We support 4.0 and 4.1 in the app, so if changing the dependency to 4.1 fixes it for 24.04, then we can do that. But we still have to support 4.0 for Ubuntu 20.04.

The newer versions we can't use, because they are for Gtk4, see https://github.com/Ulauncher/Ulauncher/blob/v6/ulauncher/utils/WebKit2.py

**Comment by oklopfer** [comment]
> That's their distros problem that they created, and only they can fix. They will have to. There are probably hundreds of apps depending on this, including many core Gnome Desktop components, [Evolution](https://gitlab.gnome.org/GNOME/evolution/-/wikis/home), Geary, Lutris and many more.

Unfortunately, this won't just be Ubuntu, as it has been changed by Debian upstream, which will eventually trickle down to all other Debian based distros: https://metadata.ftp-master.debian.org/changelogs//main/w/webkit2gtk/webkit2gtk_2.44.1-1_changelog (`webkit2gtk (2.44.0-1)`)


> We support 4.0 and 4.1 in the app, so if changing the dependency to 4.1 fixes it for 24.04, then we can do that. But we still have to support 4.0 for Ubuntu 20.04.

I believe this would be the ideal solution. I both completely understand and agree with continued support of 20.04; are you able to split up the dependencies? You may have something a bit more in-depth in mind, but for a control file, you should just be able to set `gir1.2-webkit2-4.0 | gir1.2-webkit2-4.1`, which would default to selecting `4.0`, and only `4.1` if `4.0` is unavailable.

**Comment by friday** [comment]
> you should just be able to set `gir1.2-webkit2-4.0 | gir1.2-webkit2-4.1`, which would default to selecting `4.0`, and only `4.1` if `4.0` is unavailable.

I just pushed a commit doing that, but defaulting to 4.1. I could build and install it in our Ubuntu 20.04 docker image, so it should also run, given that the app code already should support it. But please let me know if this fixes it for you.
